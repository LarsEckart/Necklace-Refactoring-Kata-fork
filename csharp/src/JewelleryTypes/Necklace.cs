using NecklaceRefactoringKata.Enums;

namespace NecklaceRefactoringKata.JewelleryTypes
{
    public record Necklace(NecklaceType Type, Jewel Stone) : JewelleryBase(Stone)
    {
    }
}
