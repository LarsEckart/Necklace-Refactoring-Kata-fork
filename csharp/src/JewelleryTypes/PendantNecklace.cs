using NecklaceRefactoringKata.Enums;

namespace NecklaceRefactoringKata.JewelleryTypes
{
    public record PendantNecklace(NecklaceType Type, Jewel Stone, Necklace Chain, JewelleryBase Pendant) : Necklace(Type, Stone)
    {
        public override bool IsLarge()
        {
            return Chain.IsLarge() || Pendant.IsLarge();
        }
    }
}
