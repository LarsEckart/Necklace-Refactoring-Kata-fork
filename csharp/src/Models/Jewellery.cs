using NecklaceRefactoringKata.Enums;

namespace NecklaceRefactoringKata.Models
{
    public abstract record Jewellery(Jewel Stone)
    {
        public virtual bool IsRing() { return false; }
        public virtual bool IsSmall() { return false; }
        public virtual bool IsEarring() { return false; }
        public virtual bool IsNecklace() { return false; }
        public virtual bool IsLarge() { return false; }
    }
}
