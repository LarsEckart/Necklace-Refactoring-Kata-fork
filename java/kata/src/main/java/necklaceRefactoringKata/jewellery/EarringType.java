package necklaceRefactoringKata.jewellery;

public enum EarringType {
    Stud,
    Hoop,
    Drop
}
