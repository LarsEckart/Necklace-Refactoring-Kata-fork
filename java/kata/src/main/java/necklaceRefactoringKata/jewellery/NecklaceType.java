package necklaceRefactoringKata.jewellery;

public enum NecklaceType {
    Beads,
    Chain,
    Pendant,
    LongChain
}
