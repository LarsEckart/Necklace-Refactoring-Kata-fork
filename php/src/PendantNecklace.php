<?php

namespace NecklaceRefactoringKata;

class PendantNecklace extends Necklace
{
    public function __construct(private Necklace $chain, private Jewellery $pendant)
    {
    }

    public function isLarge(): bool
    {
        return $this->chain->isLarge() || $this->pendant->isLarge();
    }
}
