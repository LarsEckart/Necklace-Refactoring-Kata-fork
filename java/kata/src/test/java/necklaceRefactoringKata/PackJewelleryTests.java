package necklaceRefactoringKata;

import necklaceRefactoringKata.jewellery.Jewel;
import necklaceRefactoringKata.jewellery.Jewellery;
import necklaceRefactoringKata.jewellery.Necklace;
import necklaceRefactoringKata.jewellery.NecklaceType;
import necklaceRefactoringKata.jewellery.PendantNecklace;
import org.approvaltests.combinations.CombinationApprovals;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;

import static necklaceRefactoringKata.JewelleryFactory.*;

class PackJewelleryTests {

    @Test
    void test_in_travel_roll() {
        CombinationApprovals.verifyAllCombinations(this::packItemToTravelRoll, allRingsPendantsNecklacesEarrings());
    }

    @Test
    void test_necklaces() {
        ArrayList<Necklace> necklaces = new ArrayList<>();
        necklaces.addAll(allNecklaces());
        necklaces.addAll(allPendantNecklaces());

        CombinationApprovals.verifyAllCombinations(this::packNecklace, necklaces.toArray(new Necklace[0]));
    }

    private String packNecklace(Necklace item) {
        var storage = new JewelleryStorage();

        Packer.packNecklace(item, storage);

        return StoragePrinter.print(storage);
    }

    @Test
    void test_jewellery() {
        CombinationApprovals.verifyAllCombinations(this::packItem, allRingsPendantsNecklacesEarrings());
    }

    @Test
    void test_pendantNecklace() {
        CombinationApprovals.verifyAllCombinations(this::packPendantNecklaces, Jewel.values(), NecklaceType.values(), necklaces(), allRingsPendantsNecklacesEarrings());
    }

    private String packPendantNecklaces(Jewel jewel, NecklaceType type, Necklace chain, Jewellery pendant) {
        var item = new PendantNecklace(jewel, type, chain, pendant);
        return packItem(item);
    }

    private String packItem(Jewellery item) {
        var storage = new JewelleryStorage();

        Packer.pack(item, storage);

        return StoragePrinter.print(storage);
    }

    private String packItemToTravelRoll(Jewellery item) {
        var storage = new JewelleryStorage();
        storage.travelRoll.add(item);

        Packer.pack(item, storage);

        return StoragePrinter.print(storage);
    }
}
