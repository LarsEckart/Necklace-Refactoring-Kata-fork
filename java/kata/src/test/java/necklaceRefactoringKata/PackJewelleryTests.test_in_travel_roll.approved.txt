[Ring{stone=Plain}] => box ring compartment 
[Pendant{stone=Plain}] => box top shelf 
[Necklace{type=Beads, stone=Plain}] => tree 
[Necklace{type=Chain, stone=Plain}] => box top shelf 
[Necklace{type=Pendant, stone=Plain}] => box top shelf 
[Necklace{type=LongChain, stone=Plain}] =>  
[Earring{type=Stud, stone=Plain}] => box top shelf 
[Earring{type=Hoop, stone=Plain}] => box top shelf 
[Earring{type=Drop, stone=Plain}] => box top shelf 
[Ring{stone=Diamond}] => box ring compartment 
[Pendant{stone=Diamond}] => box top shelf 
[Necklace{type=Beads, stone=Diamond}] => safe 
[Necklace{type=Chain, stone=Diamond}] => box top shelf 
[Necklace{type=Pendant, stone=Diamond}] => box top shelf 
[Necklace{type=LongChain, stone=Diamond}] => safe 
[Earring{type=Stud, stone=Diamond}] => box top shelf 
[Earring{type=Hoop, stone=Diamond}] => box top shelf 
[Earring{type=Drop, stone=Diamond}] => box top shelf 
[Ring{stone=Pearl}] => box ring compartment 
[Pendant{stone=Pearl}] => box top shelf 
[Necklace{type=Beads, stone=Pearl}] => tree 
[Necklace{type=Chain, stone=Pearl}] => box top shelf 
[Necklace{type=Pendant, stone=Pearl}] => box top shelf 
[Necklace{type=LongChain, stone=Pearl}] =>  
[Earring{type=Stud, stone=Pearl}] => box top shelf 
[Earring{type=Hoop, stone=Pearl}] => box top shelf 
[Earring{type=Drop, stone=Pearl}] => box top shelf 
[Ring{stone=Amber}] => box ring compartment 
[Pendant{stone=Amber}] => box top shelf 
[Necklace{type=Beads, stone=Amber}] => tree 
[Necklace{type=Chain, stone=Amber}] => box top shelf 
[Necklace{type=Pendant, stone=Amber}] => box top shelf 
[Necklace{type=LongChain, stone=Amber}] =>  
[Earring{type=Stud, stone=Amber}] => box top shelf 
[Earring{type=Hoop, stone=Amber}] => box top shelf 
[Earring{type=Drop, stone=Amber}] => box top shelf 
